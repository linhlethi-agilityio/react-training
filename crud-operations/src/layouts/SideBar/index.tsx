import { memo } from 'react';
import { Avatar, Heading, Text, VStack } from '@chakra-ui/react';

// Components
import { BrandLogo, Navigation, LogoutButton } from '@components';

// Constants
import { ROUTERS } from '@constants';

// Utils
import { acronymText } from '@utils';

interface SideBarProps {
  isClosed?: boolean;
  onNavigate: (path: string) => void;
  onLogout: () => void;
}

const SideBar = ({ isClosed, onNavigate, onLogout }: SideBarProps) => {
  // TODO: Mock data for now, will update later
  const user = {
    avatarUrl: '',
    name: 'John Smith',
    rule: 'Admin',
  };

  return (
    <VStack w={isClosed ? 75 : 270} px={isClosed ? 2 : 5} py={18} bg="background.sidebar" transition="1s">
      <BrandLogo isAcronym={isClosed} onClick={() => onNavigate(ROUTERS.DASHBOARD)} />

      {/* User profile */}
      <VStack boxSize="100%" spacing={20}>
        <VStack mt={54} maxW={130} mx="auto" spacing={isClosed ? 2 : 5}>
          <Avatar boxSize={isClosed ? 45 : 128} borderWidth={1} borderStyle="solid" transition="1s" />
          <VStack textAlign="center" spacing={isClosed ? 0 : 2.5}>
            <Heading fontWeight="bold" fontSize={isClosed ? 'xs' : 'sm'} lineHeight="md">
              {isClosed ? acronymText(user.name) : user.name}
            </Heading>
            <Text color="primary" fontSize={isClosed ? 'xs' : 'sm'}>
              {user.rule}
            </Text>
          </VStack>
        </VStack>

        <VStack w="100%" justifyContent="space-between" h="100%">
          {/* Menu */}
          <Navigation isShorter={isClosed} />

          {/* Logout button */}
          <LogoutButton isShorter={isClosed} onLogout={onLogout} />
        </VStack>
      </VStack>
    </VStack>
  );
};

export default memo(SideBar);
