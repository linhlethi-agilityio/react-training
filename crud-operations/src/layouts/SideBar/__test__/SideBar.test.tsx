import { fireEvent, render } from '@testing-library/react';

// Components
import SideBar from '..';

// Mocking dependencies
jest.mock('@hooks', () => ({
  useAuth: () => ({
    getCurrentUser: jest.fn(),
    logout: jest.fn(),
  }),
  useToastCustom: () => jest.fn(),
}));

jest.mock('jose', () => ({}));

jest.mock('@constants', () => ({
  ENVS: {
    VITE_API_ENDPOINT: 'import.meta.env.VITE_API_ENDPOINT',
    VITE_SECRET_KEY: 'import.meta.env.VITE_SECRET_KEY',
  },
  ROUTERS: {
    LOGIN: '/login',
  },
  SIDEBAR_NAVIGATION: [
    {
      label: 'Home',
      router: '/',
    },
    {
      label: 'Course',
      router: '/course',
    },
  ],
  BRAND_NAME: 'crud operations',
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useLocation: () => jest.fn(),
}));

const mockOnNavigate = jest.fn();

const props = {
  onLogout: jest.fn(),
  onNavigate: mockOnNavigate,
};

describe('SideBar Component', () => {
  test('renders correctly', () => {
    const { container } = render(<SideBar {...props} />);

    expect(container).toMatchSnapshot();
  });

  test('renders correctly with isClosed', () => {
    const { container } = render(<SideBar isClosed onLogout={jest.fn()} onNavigate={jest.fn()} />);

    expect(container).toMatchSnapshot();
  });

  test('call onNavigate func when click brand logo', () => {
    const { getByTestId } = render(<SideBar {...props} />);

    fireEvent.click(getByTestId('brand-logo'));

    expect(mockOnNavigate).toHaveBeenCalled();
  });
});
