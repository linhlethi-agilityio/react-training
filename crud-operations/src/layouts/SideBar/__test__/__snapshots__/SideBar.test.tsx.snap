// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar Component renders correctly 1`] = `
<div>
  <div
    class="chakra-stack css-13hpm2o"
  >
    <button
      class="chakra-link css-1mgm4dv"
      data-testid="brand-logo"
    >
      <h2
        class="chakra-heading css-qis5cd"
      >
        crud operations
      </h2>
    </button>
    <div
      class="chakra-stack css-1w47v7o"
    >
      <div
        class="chakra-stack css-a2ql3x"
      >
        <span
          class="chakra-avatar css-1uwaj3j"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <div
          class="chakra-stack css-b52v2e"
        >
          <h2
            class="chakra-heading css-vs99od"
          />
          <p
            class="chakra-text css-jlefen"
          />
        </div>
      </div>
      <div
        class="chakra-stack css-vvg5fj"
      >
        <div
          class="chakra-stack css-1kae02p"
        >
          <div
            class="chakra-stack css-oph79"
            data-testid="sidebar-item-Home"
          >
            <div
              class="css-dcng39"
            >
              <svg
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  stroke="currentColor"
                  stroke-width="1.5"
                >
                  <path
                    d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                    fill="none"
                    stroke-linecap="round"
                  />
                  <path
                    d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                    fill="currentColor"
                    stroke-linecap="round"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    fill="none"
                    r="11.25"
                    stroke-miterlimit="10"
                  />
                </g>
              </svg>
            </div>
            <p
              class="chakra-text css-0"
            >
              Home
            </p>
          </div>
          <div
            class="chakra-stack css-oph79"
            data-testid="sidebar-item-Course"
          >
            <div
              class="css-dcng39"
            >
              <svg
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  stroke="currentColor"
                  stroke-width="1.5"
                >
                  <path
                    d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                    fill="none"
                    stroke-linecap="round"
                  />
                  <path
                    d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                    fill="currentColor"
                    stroke-linecap="round"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    fill="none"
                    r="11.25"
                    stroke-miterlimit="10"
                  />
                </g>
              </svg>
            </div>
            <p
              class="chakra-text css-0"
            >
              Course
            </p>
          </div>
        </div>
        <button
          aria-label="logout"
          class="chakra-button css-4xx2wk"
          data-testid="logout-button"
          type="button"
        >
          Logout
          <span
            class="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-1jyhgcw"
              focusable="false"
              height="13"
              viewBox="0 0 17 13"
              width="17"
            >
              <path
                d="M5.313 5.208c0-.292.239-.53.53-.53h4.782V1.558c0-.235.286-.355.452-.189l4.701 4.751c.21.21.21.545 0 .754l-4.701 4.752a.265.265 0 0 1-.452-.19V8.32H5.844a.533.533 0 0 1-.532-.532v-2.58Zm-1.063 0v2.58c0 .88.714 1.594 1.594 1.594h3.718v2.055c0 1.179 1.428 1.777 2.265.94l4.705-4.748a1.595 1.595 0 0 0 0-2.258L11.827.62c-.833-.834-2.264-.243-2.264.94v2.055h-3.72c-.88 0-1.593.717-1.593 1.593ZM0 1.718v9.563c0 .88.714 1.594 1.594 1.594h4.383a.4.4 0 0 0 .398-.398v-.266a.4.4 0 0 0-.398-.399H1.594a.533.533 0 0 1-.532-.53V1.718c0-.292.24-.532.532-.532h4.383A.4.4 0 0 0 6.375.79V.523a.4.4 0 0 0-.398-.398H1.594C.714.125 0 .839 0 1.719Z"
                fill="#000000"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SideBar Component renders correctly with isClosed 1`] = `
<div>
  <div
    class="chakra-stack css-o2jq8c"
  >
    <button
      class="chakra-link css-1mgm4dv"
      data-testid="brand-logo"
    >
      <h2
        class="chakra-heading css-qis5cd"
      >
        co
      </h2>
    </button>
    <div
      class="chakra-stack css-1w47v7o"
    >
      <div
        class="chakra-stack css-1etgy3s"
      >
        <span
          class="chakra-avatar css-1hd9g3d"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <div
          class="chakra-stack css-1nq74a3"
        >
          <h2
            class="chakra-heading css-1ddqvh7"
          >
            undefined
          </h2>
          <p
            class="chakra-text css-1h928e3"
          />
        </div>
      </div>
      <div
        class="chakra-stack css-vvg5fj"
      >
        <div
          class="chakra-stack css-10i4byr"
        >
          <div
            class="chakra-stack css-1hcl0ll"
            data-testid="sidebar-item-false"
          >
            <div
              class="css-sktdp4"
            >
              <svg
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  stroke="currentColor"
                  stroke-width="1.5"
                >
                  <path
                    d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                    fill="none"
                    stroke-linecap="round"
                  />
                  <path
                    d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                    fill="currentColor"
                    stroke-linecap="round"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    fill="none"
                    r="11.25"
                    stroke-miterlimit="10"
                  />
                </g>
              </svg>
            </div>
          </div>
          <div
            class="chakra-stack css-1hcl0ll"
            data-testid="sidebar-item-false"
          >
            <div
              class="css-sktdp4"
            >
              <svg
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  stroke="currentColor"
                  stroke-width="1.5"
                >
                  <path
                    d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                    fill="none"
                    stroke-linecap="round"
                  />
                  <path
                    d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                    fill="currentColor"
                    stroke-linecap="round"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    fill="none"
                    r="11.25"
                    stroke-miterlimit="10"
                  />
                </g>
              </svg>
            </div>
          </div>
        </div>
        <button
          aria-label="logout"
          class="chakra-button css-4xx2wk"
          data-testid="logout-button"
          type="button"
        >
          <span
            class="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-1uwn6wz"
              focusable="false"
              height="13"
              viewBox="0 0 17 13"
              width="17"
            >
              <path
                d="M5.313 5.208c0-.292.239-.53.53-.53h4.782V1.558c0-.235.286-.355.452-.189l4.701 4.751c.21.21.21.545 0 .754l-4.701 4.752a.265.265 0 0 1-.452-.19V8.32H5.844a.533.533 0 0 1-.532-.532v-2.58Zm-1.063 0v2.58c0 .88.714 1.594 1.594 1.594h3.718v2.055c0 1.179 1.428 1.777 2.265.94l4.705-4.748a1.595 1.595 0 0 0 0-2.258L11.827.62c-.833-.834-2.264-.243-2.264.94v2.055h-3.72c-.88 0-1.593.717-1.593 1.593ZM0 1.718v9.563c0 .88.714 1.594 1.594 1.594h4.383a.4.4 0 0 0 .398-.398v-.266a.4.4 0 0 0-.398-.399H1.594a.533.533 0 0 1-.532-.53V1.718c0-.292.24-.532.532-.532h4.383A.4.4 0 0 0 6.375.79V.523a.4.4 0 0 0-.398-.398H1.594C.714.125 0 .839 0 1.719Z"
                fill="#000000"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
